ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -933,6 +933,28 @@ let get_SRs_required_for_recovery ~__context ~self ~session_to =
 
 
 (* BIOS strings *)
+let assert_valid_bios_strings ~__context ~bios_strings =
+  (* Validate the bios_string keys are listed in standard_bios_strings *)
+  (* Validate size of value provided is within bios_string_limit_size and not empty *)
+  (* Validate value chars are printable ASCII characters *)
+  List.iter (fun (k, v) ->
```

Comment:
This is works, but wouldn't it be more readable the way I've written it below?
And using `String.iter` is neater and more efficient than the combination of `Xstringext.String.explode` and `List.for_all`: see http://caml.inria.fr/pub/docs/manual-ocaml/libref/String.html
```
  bios_strings |> List.iter (fun (k, v) ->
    if (not List.mem_assoc k Xapi_globs.standard_bios_strings) then (
      let expected_keys = String.concat ", " (List.map (fun (a, _) -> a) Xapi_globs.standard_bios_strings) in
      raise (Api_errors.Server_error(Api_errors.invalid_key, [expected_keys; k]))
    );
    match String.length v with
      | 0 -> raise (Api_errors.Server_error(Api_errors.invalid_value, [k; "Value provided is empty"]))
      | len when len > Xapi_globs.bios_string_limit_size ->
          raise (Api_errors.Server_error(Api_errors.invalid_value, [k; v ^ " has length more than 512 characters"]))
      | _ ->
        String.iter
          (fun c ->
            if c < (Char.chr 32) || c >= (Char.chr 127) then
              raise (Api_errors.Server_error(Api_errors.invalid_value, [k; v ^ " has non-printable ASCII characters"]))
          ) v
  )
```

Reply:
I think that we should change the definition of `set_bios_strings` in the datamodel to use an enum type for the keys, so the `value` parameter has type `(enum bios_string_keys -> string) map`. Then the key name checking will be handled automatically, and we won't need the new API error. We probably don't need to split the lists in xapi_globs.ml anymore either.